### ISSUE TITLE
Generic type not assignable when it's wrapped within separate generic type alias.

### ISSUE THREAD

**Issue by ryanstaniforth** [Microsoft/TypeScript#25177]
**TypeScript Version:**  3.0.0-dev.20180623

**Search Terms:** generic extends not assignable

**Code**

```ts
type InputHandler<MessageInterface> = (input: MessageInterface) => void;

interface A {
    a: number;
}

// Example 1: does work
const handlers: A[] = [];

const func = <MessageInterface extends A>(handler: MessageInterface): void => {
    handlers.push(handler);
};

// Example 2: doesn't work
const handlers2: Array<InputHandler<A>> = [];

const func2 = <MessageInterface extends A>(handler: InputHandler<MessageInterface>): void => {
    handlers2.push(handler); // Error here
};
```

**Expected behavior:** No errors

**Actual behavior:** In the second example I get the following error:
```
error TS2345: Argument of type 'InputHandler<MessageInterface>' is not assignable to parameter of type 'InputHandler<A>'.
  Types of parameters 'input' and 'input' are incompatible.
    Type 'A' is not assignable to type 'MessageInterface'.
```
This error only appears with `--strictFunctionTypes`, but I still don't understand why it's not assignable.

**Playground Link:** [Link](http://www.typescriptlang.org/play/#src=export%20type%20InputHandler%3CMessageInterface%3E%20%3D%20(input%3A%20MessageInterface)%20%3D%3E%20void%3B%0D%0A%0D%0Ainterface%20A%20%7B%0D%0A%20%20%20%20a%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aconst%20handlers%3A%20A%5B%5D%20%3D%20%5B%5D%3B%0D%0A%0D%0Aexport%20const%20func%20%3D%20%3CMessageInterface%20extends%20A%3E(handler%3A%20MessageInterface)%3A%20void%20%3D%3E%20%7B%0D%0A%20%20%20%20handlers.push(handler)%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20handlers2%3A%20Array%3CInputHandler%3CA%3E%3E%20%3D%20%5B%5D%3B%0D%0A%0D%0Aexport%20const%20func2%20%3D%20%3CMessageInterface%20extends%20A%3E(handler%3A%20InputHandler%3CMessageInterface%3E)%3A%20void%20%3D%3E%20%7B%0D%0A%20%20%20%20handlers2.push(handler)%3B%0D%0A%7D%3B%0D%0A)

**Comment by jack-williams** [comment]
Here is unsound behaviour you can get otherwise:

```ts
interface B {
    a: number;
    b: number;
}

func2((b: B) => b.a + b.b);
handlers2[0]({ a: 3 }); // oops! The object doesn't have property b.
```

The generic argument in `func2` allows you to supply an argument that is more specific than `A` (requires more information), but will be exposed at the less specific type of `A` in `handlers2`. A caller will provide an argument with the information of type `A`, but this might be insufficient for the type you picked when calling `func2`.

**Comment by jack-williams** [comment]
The difference in `handlers` and `handlers2` is that in the former A appears in a covariant position, and in the latter it appears in a contravariant position. Your generic parameter is a least as specific as `A`, but it could be more specific. 

This is ok for covariance: `MessageInterface is a subtype of A`,
but not for contravariance: `A is not always a subtype of MessageInterface`.

Consider the unsound example:
```ts
interface B {
    a: number;
    b: number;
}

func2((b: B) => b.a + b.b);
handlers2[0]({ a: 3 }); // oops! The object doesn't have property b.
```

**Comment by DanielRosenwasser** [comment]
Thanks @jack-williams!
